HPCC-31496 Allow field translation that only removes fields - and does not add blank values

jackdelv opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jackdelv committed:
HPCC-31496 Allow field translation that only removes fields - and does not add blank values

jackdelv commented:
@ghalliday Is this the behaviour that you were imagining? Is this the best place to add the PayloadRemoveOnly logic?

richardkchapman commented on common/thorhelper/thorcommon.cpp:
```diff
@@ -2123,6 +2123,15 @@ static bool getTranslators(Owned<const IDynamicTransform> &translator, Owned<con
+            if (publishedCrc && expectedCrc && (publishedCrc != expectedCrc))
+            {
+                DBGLOG("Overriding stored record layout reading file %s", tracing);
+                if (expectedFormat->queryRecordAccessor(true).getNumFields() > publishedFormat->queryRecordAccessor(true).getNumFields())
```
I'm not sure if this is supposed to check that fields were only removed and not added, but how would it happen a case where fields were both added and removed (but more were removed than added)?

jackdelv replied:
I fixed the check to use match_none. Now it strictly checks for added fields.

richardkchapman commented on roxie/ccd/ccdfile.cpp:
```diff
@@ -2965,6 +2965,15 @@ class CResolvedFile : implements IResolvedFileCreator, implements ISafeSDSSubscr
+                        if (formatCrcs.item(idx) && expectedFormatCrc && (formatCrcs.item(idx) != expectedFormatCrc))
+                        {
+                            DBGLOG("Overriding stored record layout reading file %s", subname);
+                            if (expected->queryRecordAccessor(true).getNumFields() > diskTypeInfo.item(idx)->queryRecordAccessor(true).getNumFields())
```
Same comment as before.

richardkchapman commented on rtl/eclrtl/rtldynfield.hpp:
```diff
@@ -110,7 +110,8 @@ enum class RecordTranslationMode : byte
     AlwaysECL = 4,  // Ignore the published format - can make sense to force no translation e.g. when field names have changed
-    Unspecified = 5
+    Unspecified = 5,
+    PayloadRemoveOnly = 6 // Allow fields to be removed from the incoming dataset, but not allow fields to be missing
```
I would have made Unspecified be 6... these values are not persisted anywhere AFAIK

jackdelv replied:
Changed Unspecified to 6.

richardkchapman reviewed (dismissed):
There are other references to RecordTranslationMode::Payload that you might want to change, if the intention was to change the default.

But I don't think as coded this actually works as intended.

ghalliday commented:
@jackdelv following on from Richard's comments.  You need to add a method to IDynamicTransform e.g. hasNewFields().  Take a look at match_none in FieldMatchType which should be set if that is the case.

jackdelv commented:
@richardkchapman @ghalliday I addressed your comments and made appropriate changes. However, I am uncertain how to verify that I have covered all the cases where hasNewFields needs to be called. I looked in the repo for uses of "RecordTranslationMode::Payload", "payload", "needsTranslate". Should there be a check for added fields everywhere needsTranslate it used?

ghalliday commented on helm/hpcc/values.schema.json:
```diff
@@ -1825,8 +1825,8 @@
-          "default": "payload",
-          "enum": ["false", "true", "payload"],
+          "default": "PayloadRemoveOnly",
+          "enum": ["false", "true", "PayloadRemoveOnly"],
```
payload still needs to be in the list as a valid option.  It would be better to have consistent casing with the other options - a leading lower case letter.

ghalliday reviewed (commented):
This looks like it is close to being ready
- needs to be a non-draft PR
- I think it is missing a check in slavmain.cpp(~355)
- Missing some places in hthorkey.cpp (search for canTranslate())

I am not quite sure what version to target.  Probably 9.6.x.  An alternative, is to target the change to 9.4.x, by the change in defaults to 9.6..x.
@jakesmith thoughts?

jackdelv commented:
@ghalliday I add payload back as an option and changed the casing to be more consistent. I also added checks in slavmain.cpp and hthorkey.cpp. Back to you.

jakesmith commented on common/thorhelper/thorread.cpp:
```diff
@@ -176,6 +176,9 @@ void DiskReadMapping::ensureTranslators() const
             throw MakeStringException(0, "Untranslatable record layout mismatch detected for file %s", filename);
 
+        if (mode == RecordTranslationMode::PayloadRemoveOnly && translator->hasNewFields())
+                throw MakeStringException(0, "Untranslatable record layout mismatch detected for file %s. Expected fields were not found in source.", filename);
```
formatting: indentation off.

jackdelv replied:
Fixed

jakesmith commented on ecl/hthor/hthorkey.cpp:
```diff
@@ -700,6 +700,8 @@ const IDynamicTransform * CHThorIndexReadActivityBase::getLayoutTranslator(IDist
     if (!payloadTranslator->canTranslate())
         throw MakeStringException(0, "Untranslatable key layout mismatch reading index %s", f->queryLogicalName());
+    if (getLayoutTranslationMode() == RecordTranslationMode::PayloadRemoveOnly && payloadTranslator->hasNewFields())
+                throw MakeStringException(0, "Untranslatable record layout mismatch detected for file %s. Expected fields were not found in source.", f->queryLogicalName());
```
formatting: indentation off.

jackdelv replied:
Fixed

jakesmith commented on common/thorhelper/thorcommon.cpp:
```diff
@@ -2148,6 +2148,9 @@ static bool getTranslators(Owned<const IDynamicTransform> &translator, Owned<con
                 throw MakeStringException(0, "Untranslatable record layout mismatch detected for file %s", tracing);
 
+            if (mode == RecordTranslationMode::PayloadRemoveOnly && translator->hasNewFields())
+                throw MakeStringException(0, "Untranslatable record layout mismatch detected for file %s. Expected fields were not found in source.", tracing);
```
minor: It is translatable, but mode prohibits it. Consider different message here (and similarly elsewhere).

jackdelv replied:
Changed message.

jakesmith reviewed (commented):
@jackdelv - added a couple of minor comments.

jakesmith commented on initfiles/etc/DIR_NAME/environment.xml.in:
```diff
@@ -68,7 +68,7 @@
                 enableSysLog="true"
                 fastLaneQueue="true"
-                fieldTranslationEnabled="payload"
+                fieldTranslationEnabled="payloadRemoveOnly"
```
This should be consisten with testing/regress/environment.xml.in.

jackdelv replied:
The changes between initfiles/etc/DIR_NAME/environment.xml.in and testing/regress/environment.xml.in look consistent to me. I'm probably not misunderstanding what you mean.

jakesmith replied:
agreed, I missed the testing/regress change.

jakesmith commented:
>I am not quite sure what version to target. Probably 9.6.x. An alternative, is to target the change to 9.4.x, by the change in defaults to 9.6..x.
@jakesmith thoughts?

I think it's safe enough to go into 9.4 if we leave the default as 'payload'. And if many are going to stick with 9.4 for a while, and want this feature, they'll have it there. So I'd vote for this targeting 9.4

ghalliday reviewed (approved):
Looks good to me

ghalliday commented:
@jackdelv the regression tests will need to be updated to set the option in the work unit to override the default

jackdelv commented:
@ghalliday There are still some regression tests failing like sqtrans.ecl that hadn't previously set a value for the translationMode. Is overriding the default in the ecl files correct, or did you mean something different?

ghalliday commented:
@jackdelv yes, please add a similar #option into those files to force the old payload option on.

jackdelv commented:
@ghalliday I believe all the necessary regression tests have been fixed.

jakesmith reviewed (approved):
@jackdelv - looks good.

ghalliday reviewed (approved):
Looks good.  Just re-running k8s tests.
Also debating whether this should go in 9.6.x.  It is a bit late - because it will potentially cause existing queries to fail, which shouldn't happen in a point release.  However, without it there is no safe way of enabiling field translation if you do not want missing fields to be allowed.

Possibly one way forward would be to merge it to 9.6.x with payload as the default, and then change the default for master/9.8.x.

ghalliday commented:
Please squash, and lets talk about the target build.

jackdelv commented:
@ghalliday Squashed.

ghalliday reviewed (commented):
@jackdelv please can you split this PR in two.
One set of changes that targets 9.6.x which adds the code to process everything, but does not change the default.  Then a second PR targeting master that changes the default.
